Eclipse plugin

oleludwig opened the pull request:
<!-- This PR fixes #NUMBER_OF_THE_ISSUE, and fixes #NUMBER_OF_THE_ISSUE -->

## Description

This PR adds an eclipse plugin for the sapl policy language. It uses the support from eclipse support from xtext (https://eclipse.dev/Xtext/documentation/310_eclipse_support.html).
The plugin provides features like syntax highlighting, content assist, an outline, folding and shows errors in the policy file.

The PR adds a project which generates an eclipse update site in eclipse-plugin/sapl-eclipse-repository/target/repository which can be manually added to eclipse. The plugin can be installed from this update site. I tested the installation and usage of the plugin under linux and windows.

Out of scope:
- This PR only includes a plugin for sapl policy files not for sapl-test. Support for sapl-test can be added in a follow-up PR.

How to test:

- Build the project with `mvn clean install` locally.
- Open the "Install" dialog in eclipse (Help -> Install New Software...).
- Add the local update-site  to eclipse (Click on "Add..." , then on "Local..." and select the local update-site. Its in the local git repository in the subfolder "eclipse-plugin/sapl-eclipse-repository/target/repository")
- Select the update-site in the dropdown after "Work with". Its name is "SAPL Eclipse update site".
- Select "SAPL in the table below.
- Click on "Next".
- Click on "Finish".
- Trust all bundles for the sapl plugin in the "Trust Artifacts" dialog. ("Select All" and then "Trust Selected")
- Restart eclipse from the "Software Updates" dialog. ("Restart Now")
- Open a sapl file.

Known problems which can be fixed later and should not block this PR:
- auto formatting isn't using the preferred formatting style yet.
- outline view contains some unnamed nodes.

## Checklist

<!--
Fill out and perform the following checklist before publishing the PR for review.
-->

* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.
* [X] All new and existing tests passed.
* [X] My code follows the code style of this project.
* [X] I ran all checks which produced no new errors or warnings for my changes.
* [X] I have checked to ensure there aren't other open Pull Requests for the same update/change.

<!-- 📛📛📛
All pull requests go through pipelines and careful peer review to ensure the quality of the code, but this does not exempt developers from delivering good code. 
Write code to the best of your ability while considering best practices, guidelines, and requirements.

If it fixes any current issues please let us know this way:
Uncomment the comment above the "description", then add your number of issues after the "#".
Example: # **This pull request fixes #NUMBER_OF_THE_ISSUE issue**
If there are multiple issues to be closed with the merge of this pull request
please do it like so: **This pull request fixes #NUMBER_OF_THE_ISSUE, fixes #NUMBER_OF_THE_ISSUE and fixes #NUMBER_OF_THE_ISSUE issue**.
For more information on closing issues using keywords, please check https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords#closing-multiple-issues
📛📛📛 -->

oleludwig commented:
@heutelbeck this PR is ready for review from my perspective. There are some know minor issues (see description above) which I will resolve in follow-up PRs. I don't know why I can't see the SonarCloud analysis in this PR. As we can see in PR #149 the sonar cloud analysis should already have been visible here when this PR was a draft. I also can't find a result of the analysis when searching in SonarCloud directly.

heutelbeck commented on eclipse-plugin/sapl-eclipse-feature/feature.xml:
```diff
@@ -0,0 +1,42 @@
+<feature id="sapl-eclipse-feature"
+	label="SAPL Feature "
+	version="3.0.0.qualifier">
+<!--	<plugin-->
```
Delete commented code?

oleludwig replied:
Removed now. It was left over from an earlier state of the eclipse plugin before the restructuring where I converted sapl-lang and sapl-ide from eclipse-plugins to regular osgi-bundles.

heutelbeck commented on eclipse-plugin/sapl-eclipse-ui/src/main/java/io/sapl/grammar/ui/labeling/SAPLDescriptionLabelProvider.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class SAPLDescriptionLabelProvider extends DefaultDescriptionLabelProvider {
+
+    // Labels and icons can be computed like this:
```
Commented out code. Here I understand that this is a TODO to implement providing icons to Eclipse. Is this on the menu?

oleludwig replied:
From my understanding this class is for providing custom icons and labels for the outline and some other places.
I do not plan to add custom icons to the outline at the moment. The default implementation here might be enough, even if we decide to add some custom icons later.

I will try to change the icon for sapl policy files to the sapl icon, but this probable needs to be changed in the plugin.xml instead of here. (see https://www.eclipse.org/forums/index.php/t/1109080/)

This is probably the right place to customize labels in the outline and may be needed to fix one of the known issues. But I will try to fix the missing names in the outline in sapl-lang and sapl-ide first.

heutelbeck commented on eclipse-plugin/sapl-eclipse-ui/src/main/java/io/sapl/grammar/ui/quickfix/SAPLQuickfixProvider.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class SAPLQuickfixProvider extends DefaultQuickfixProvider {
+
+    // @Fix(SAPLValidator.INVALID_NAME)
```
Same here. 

Proposal for implementation: Lazy boolean operators to be replaced by eager ones where applicable. I can explain in detail. DM me.

oleludwig replied:
Removed the example code for now. I will contact you regarding your proposal after fixing the remaining issues.

heutelbeck reviewed (approved):
Will approve. Still, take care of the code in comments. We need to see whats wrong with Sonar anyway and react to the analysis when it comes in.

oleludwig commented:
I addressed the comments and updated the PR. I'm not sure why the PR wasn't merged despite being approved by @heutelbeck . Maybe there is some additional confirmation required?

heutelbeck commented:
> I addressed the comments and updated the PR. I'm not sure why the PR wasn't merged despite being approved by @heutelbeck . Maybe there is some additional confirmation required?

I manually triggered the SonarCloud analysis and wanted to take a look at that first.

oleludwig commented:
@heutelbeck I added a unit test for the sapl-eclipse-ui project and fixed the warnings in the build as far as possible.
The project requires maven 3.9 now to be built due to the upgrade to tycho 4.

Regarding the failed quality gate in sonarcloud:
- The missing code coverage only consists of empty classes, constructors which only call super and the bind-Methods in the SAPLUiModule which only reference the other classes. I don't see the benefit of adding tests for this, but can add some just for meeting the coverage goal, if you prefer.
- The two issues from sonarcloud are caused by wrapping the InitializationException into a RuntimeException in the SAPLUiContentProposalProvider. First of all the InitializationException can't happen here at runtime. This is also verified by the unit test. I also can't throw the exception much further because we are restricted by the interface of the IdeContentProposalProvider. I also see no proper way of handling this Exception here. A SneakyThrows would probably be appropiate, but it isn't worth the effort to integrate lombok with tycho just for one SneakyThrows. So I suggest to ignore these issues. I can add a custom runtime exception instead of the generic one, if you prefer.

heutelbeck commented:
Thank you, Ole.

* Coverage. Yes, please add the tests. Even though they seem useless now, they may be able to catch regression issues in the future.
* Exceptions. I suggest to add ```@SneakyThrows``` instead of the try/catch block.

heutelbeck commented:
Just rereading the code regarding the exceptions.
Why not just do the initialization in a constructor that then may throw ?
Doing all the "new ..." in the method seems to introduce a lot of hard coupling.
I would prefer a constructor injection of the contexts, so that these can be exchanged in the future. Which is more likely than you may expect at the moment.

oleludwig commented:
I refactored the SAPLUiContentProposalProvider to use dependency injection to get the PDPConfigurationProvider like we discussed. The PDPConfigurationProvider is now provided by a bind method in SAPLUiModule. Please let me know if you prefer to break this further down into smaller bind methods.

The test coverage is also fixed now.

I will try to combine the SAPLUiContentProposalProvider and the SAPLContentProposalProvider in the next days so that we don't need the SAPLUiContentProposalProvider and maybe the SpringContext anymore. But we don't necessarily need to block this PR for it.
